File: internal/browser/browser_test.go
package browser

import (
	"testing"
	"time"
)

func unixToChromeExpires(unix int64) int64 {
	return (unix + 11_644_473_600) * 1_000_000
}

func TestPKCS7Unpad(t *testing.T) {
	// 合法 PKCS7 padding 应被正确去除。
	valid := append([]byte("DATA"), []byte{4, 4, 4, 4}...)
	out, err := pkcs7Unpad(valid)
	if err != nil {
		t.Fatalf("valid unpad should not error: %v", err)
	}
	if string(out) != "DATA" {
		t.Fatalf("unexpected unpad output: %q", string(out))
	}

	// 非法 padding 应返回错误。
	_, err = pkcs7Unpad([]byte{1, 2, 3, 0})
	if err == nil {
		t.Fatalf("invalid padding should error")
	}
}

func TestHostMatches(t *testing.T) {
	cases := []struct {
		note   string
		target string
		cookie string
		want   bool
	}{
		{"测子域命中；预期 true", "a.example.com", ".example.com", true},
		{"测根域命中；预期 true", "example.com", ".example.com", true},
		{"测完全相等域名；预期 true", "example.com", "example.com", true},
		{"测无关域名；预期 false", "bad.com", ".example.com", false},
	}
	for _, tc := range cases {
		// 参考 table note。
		if got := hostMatches(tc.target, tc.cookie); got != tc.want {
			t.Fatalf("hostMatches(%q,%q)=%v want %v", tc.target, tc.cookie, got, tc.want)
		}
	}
}

func TestNormalizePath(t *testing.T) {
	// 空路径归一化为根路径。
	if normalizePath("") != "/" {
		t.Fatalf("empty path should normalize to /")
	}
	// 缺少末尾 / 时自动补全。
	if normalizePath("/zentao") != "/zentao/" {
		t.Fatalf("missing trailing slash should be appended")
	}
	// 已有末尾 / 不应被改变。
	if normalizePath("/zentao/") != "/zentao/" {
		t.Fatalf("existing trailing slash should remain")
	}
}

func TestChromeExpiresUTCToUnix(t *testing.T) {
	// 0 代表 session，转换后仍为 0。
	if chromeExpiresUTCToUnix(0) != 0 {
		t.Fatalf("zero expires should map to zero unix")
	}
	// Chrome epoch 精确转换为 Unix epoch。
	if got := chromeExpiresUTCToUnix(11_644_473_600 * 1_000_000); got != 0 {
		t.Fatalf("chrome epoch mismatch: got %d", got)
	}
}

func TestChooseBestByPath(t *testing.T) {
	// 同名 cookie 选择“未过期 + path 最长”的项。
	future := unixToChromeExpires(time.Now().Add(24 * time.Hour).Unix())
	past := unixToChromeExpires(time.Now().Add(-24 * time.Hour).Unix())

	items := []BrowserCookieItem{
		{Name: "zp", Value: "expired", Path: "/zentao/deep/", ExpiresUTC: past},
		{Name: "zp", Value: "root", Path: "/", ExpiresUTC: future},
		{Name: "zp", Value: "deep", Path: "/zentao/", ExpiresUTC: future},
		{Name: "za", Value: "other", Path: "/zentao/", ExpiresUTC: future},
	}
	best := chooseBestByPath(items, "zp")
	if best == nil {
		t.Fatalf("expected best item")
	}
	if best.Value != "deep" {
		t.Fatalf("expected deepest non-expired path, got %q", best.Value)
	}
}

File: src/browser_test.rs
use super::*;

// Unix 时间戳转 Chrome epoch 微秒，便于断言转换逻辑。
fn unix_to_chrome_expires(unix: i64) -> i64 {
    (unix + 11_644_473_600) * 1_000_000
}

// 覆盖根域、子域命中与不命中三种情况。
#[test]
fn test_host_matches() {
    assert!(host_matches("a.example.com", ".example.com"));
    assert!(host_matches("example.com", "example.com"));
    assert!(!host_matches("bad.com", ".example.com"));
}

// 路径归一化应保证末尾带 /，空路径归一为 /。
#[test]
fn test_normalize_path() {
    assert_eq!(normalize_path(""), "/");
    assert_eq!(normalize_path("/zentao"), "/zentao/");
    assert_eq!(normalize_path("/zentao/"), "/zentao/");
}

// Chrome epoch 与 Unix epoch 转换结果应可预期且稳定。
#[test]
fn test_expires_convert() {
    assert_eq!(chrome_expires_utc_to_unix(0), 0);
    let expires = unix_to_chrome_expires(1704067200);
    assert_eq!(chrome_expires_utc_to_unix(expires), 1704067200);
}

// 同名 cookie 应选择“未过期 + path 最长”的项。
#[test]
fn test_choose_best_by_path() {
    let future = unix_to_chrome_expires(
        (std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .expect("time should be valid")
            .as_secs() as i64)
            + 24 * 3600,
    );
    let past = unix_to_chrome_expires(
        (std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .expect("time should be valid")
            .as_secs() as i64)
            - 24 * 3600,
    );

    let items = vec![
        BrowserCookieItem {
            name: "zp".to_string(),
            value: "expired".to_string(),
            domain: "example.com".to_string(),
            path: "/zentao/deep/".to_string(),
            secure: false,
            http_only: true,
            expires_utc: past,
            creation_utc: 10,
            last_access_utc: 20,
        },
        BrowserCookieItem {
            name: "zp".to_string(),
            value: "root".to_string(),
            domain: "example.com".to_string(),
            path: "/".to_string(),
            secure: false,
            http_only: true,
            expires_utc: future,
            creation_utc: 30,
            last_access_utc: 40,
        },
        BrowserCookieItem {
            name: "zp".to_string(),
            value: "deep".to_string(),
            domain: "example.com".to_string(),
            path: "/zentao/".to_string(),
            secure: false,
            http_only: true,
            expires_utc: future,
            creation_utc: 50,
            last_access_utc: 60,
        },
    ];

    let best = choose_best_by_path(&items, "zp").expect("best cookie should exist");
    assert_eq!(best.value, "deep");
}

// 同 path 的多条 cookie 应选择最近访问（重新登录后最新）的记录。
#[test]
fn test_choose_best_by_latest_access_when_path_equal() {
    let future = unix_to_chrome_expires(
        (std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .expect("time should be valid")
            .as_secs() as i64)
            + 24 * 3600,
    );

    let items = vec![
        BrowserCookieItem {
            name: "zp".to_string(),
            value: "old-session".to_string(),
            domain: "example.com".to_string(),
            path: "/zentao/".to_string(),
            secure: false,
            http_only: true,
            expires_utc: future,
            creation_utc: 100,
            last_access_utc: 100,
        },
        BrowserCookieItem {
            name: "zp".to_string(),
            value: "new-session".to_string(),
            domain: "example.com".to_string(),
            path: "/zentao/".to_string(),
            secure: false,
            http_only: true,
            expires_utc: future,
            creation_utc: 200,
            last_access_utc: 200,
        },
    ];

    let best = choose_best_by_path(&items, "zp").expect("best cookie should exist");
    assert_eq!(best.value, "new-session");
}
